### DISCUSSION TITLE
Getting roots from tree sequence

### DISCUSSION THREAD

**Discussion by haono** [tskit-dev/tskit#2299] [Q&A] [answered]
Hello,
Is there any way to obtain all the roots from a tree sequence without iterating over single trees?
Many thanks.

**Comment by benjeffery** [comment, marked as answer]
Hi @haono!
Thanks for your question. The TreeSequence doesn't store any cached information about roots, so to find them you have to work through the trees. Are you finding this to be a bottleneck? It shouldn't be that slow to go through the trees. Feel free to paste some code of how you're doing this if you want us to have a look.

**Reply by haono** [reply]
Thank you very much @benjeffery

**Comment by jeromekelleher** [comment]
In special cases you might be able to look at the oldest edges and reason about them, but it would be quite tricky I think. As @benjeffery says though, going through the trees should be quite a quick way to do this unless you've got a large number of roots.

**Reply by haono** [reply]
Thank you very much @jeromekelleher
